File: examples/python/api_usage_example.py
"""
NudgeSafe Hands - Exemple d'Utilisation API Python
Démonstration des appels API principaux
"""

import requests
import json
from typing import Dict, List, Optional

# Configuration
API_BASE_URL = "http://localhost:3000/api/v1"
API_KEY = "your_api_key_here"  # À remplacer

class NudgeSafeClient:
    """Client Python pour l'API NudgeSafe Hands"""
    
    def __init__(self, base_url: str = API_BASE_URL, api_key: str = API_KEY):
        self.base_url = base_url
        self.headers = {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {api_key}"
        }
    
    def get_glove_recommendation(
        self, 
        chemical_name: str,
        exposure_time: int = 60,
        task_type: str = "manipulation"
    ) -> Dict:
        """
        Obtenir une recommandation de gant pour un produit chimique
        
        Args:
            chemical_name: Nom du produit chimique
            exposure_time: Durée d'exposition en minutes
            task_type: Type de tâche (manipulation, cleaning, immersion)
        
        Returns:
            Dict avec recommandations de gants
        """
        endpoint = f"{self.base_url}/recommendations/glove"
        payload = {
            "chemical": chemical_name,
            "exposure_time_minutes": exposure_time,
            "task_type": task_type
        }
        
        response = requests.post(endpoint, json=payload, headers=self.headers)
        response.raise_for_status()
        return response.json()
    
    def parse_sds(self, sds_file_path: str) -> Dict:
        """
        Parser une Fiche de Données de Sécurité (FDS/SDS)
        
        Args:
            sds_file_path: Chemin vers le fichier PDF de la FDS
        
        Returns:
            Dict avec informations extraites
        """
        endpoint = f"{self.base_url}/parse/sds"
        
        with open(sds_file_path, 'rb') as f:
            files = {'file': (sds_file_path, f, 'application/pdf')}
            # Supprimer Content-Type header pour multipart
            headers = {k: v for k, v in self.headers.items() if k != "Content-Type"}
            response = requests.post(endpoint, files=files, headers=headers)
        
        response.raise_for_status()
        return response.json()
    
    def get_chemical_info(self, cas_number: Optional[str] = None, name: Optional[str] = None) -> Dict:
        """
        Obtenir les informations d'un produit chimique
        
        Args:
            cas_number: Numéro CAS du produit
            name: Nom du produit (si CAS non disponible)
        
        Returns:
            Dict avec informations produit
        """
        endpoint = f"{self.base_url}/chemicals"
        params = {}
        
        if cas_number:
            params['cas'] = cas_number
        elif name:
            params['name'] = name
        else:
            raise ValueError("CAS number or name required")
        
        response = requests.get(endpoint, params=params, headers=self.headers)
        response.raise_for_status()
        return response.json()
    
    def get_glove_materials(self) -> List[Dict]:
        """
        Obtenir la liste de tous les matériaux de gants disponibles
        
        Returns:
            List de matériaux avec propriétés
        """
        endpoint = f"{self.base_url}/gloves/materials"
        response = requests.get(endpoint, headers=self.headers)
        response.raise_for_status()
        return response.json()


# =============================================================================
# EXEMPLES D'UTILISATION
# =============================================================================

def example_basic_recommendation():
    """Exemple 1: Recommandation simple de gant"""
    client = NudgeSafeClient()
    
    # Obtenir recommandation pour acétone
    result = client.get_glove_recommendation(
        chemical_name="Acetone",
        exposure_time=30,
        task_type="cleaning"
    )
    
    print("=== Recommandation pour Acétone ===")
    print(f"Produit: {result['chemical']['name']}")
    print(f"CAS: {result['chemical']['cas_number']}")
    print("\nGants recommandés:")
    for glove in result['recommendations']['recommended']:
        print(f"  ✅ {glove['material']} - Temps percée: {glove['breakthrough_time']}min")
    
    print("\nGants NON recommandés:")
    for glove in result['recommendations']['not_recommended']:
        print(f"  ❌ {glove['material']} - Raison: {glove['reason']}")


def example_sds_parsing():
    """Exemple 2: Parser une FDS automatiquement"""
    client = NudgeSafeClient()
    
    # Parser FDS
    result = client.parse_sds("path/to/safety_data_sheet.pdf")
    
    print("=== Parsing FDS ===")
    print(f"Produit: {result['product_name']}")
    print(f"CAS: {result['cas_number']}")
    print(f"Section 8 - EPI:")
    print(f"  Gants: {', '.join(result['ppe']['gloves'])}")
    print(f"  Protection respiratoire: {result['ppe']['respiratory']}")


def example_chemical_lookup():
    """Exemple 3: Recherche produit chimique par CAS"""
    client = NudgeSafeClient()
    
    # Rechercher par CAS
    result = client.get_chemical_info(cas_number="67-64-1")
    
    print("=== Info Produit Chimique ===")
    print(f"Nom: {result['name']}")
    print(f"CAS: {result['cas_number']}")
    print(f"Famille: {result['family']}")
    print(f"Dangers:")
    for hazard in result['hazards']:
        print(f"  - {hazard}")


def example_glove_materials():
    """Exemple 4: Lister tous les matériaux de gants"""
    client = NudgeSafeClient()
    
    materials = client.get_glove_materials()
    
    print("=== Matériaux de Gants Disponibles ===")
    for material in materials:
        print(f"\n{material['name']}:")
        print(f"  Résistance chimique: {material['chemical_resistance']}/5")
        print(f"  Résistance mécanique: {material['mechanical_resistance']}/5")
        print(f"  Dextérité: {material['dexterity']}/5")
        print(f"  Prix: {'$' * material['cost_level']}")


if __name__ == "__main__":
    print("NudgeSafe Hands - Exemples API Python\n")
    
    try:
        example_basic_recommendation()
        print("\n" + "="*60 + "\n")
        
        # Décommentez pour tester les autres exemples
        # example_sds_parsing()
        # example_chemical_lookup()
        # example_glove_materials()
        
    except requests.exceptions.RequestException as e:
        print(f"❌ Erreur API: {e}")
    except Exception as e:
        print(f"❌ Erreur: {e}")

File: examples/javascript/api_usage_example.js
/**
 * NudgeSafe Hands - Exemple d'Utilisation API JavaScript
 * Démonstration des appels API principaux
 */

const API_BASE_URL = 'http://localhost:3000/api/v1';
const API_KEY = 'your_api_key_here'; // À remplacer

/**
 * Client JavaScript pour l'API NudgeSafe Hands
 */
class NudgeSafeClient {
  constructor(baseUrl = API_BASE_URL, apiKey = API_KEY) {
    this.baseUrl = baseUrl;
    this.apiKey = apiKey;
  }

  /**
   * Headers par défaut pour les requêtes
   */
  get headers() {
    return {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${this.apiKey}`,
    };
  }

  /**
   * Obtenir une recommandation de gant pour un produit chimique
   */
  async getGloveRecommendation(chemicalName, exposureTime = 60, taskType = 'manipulation') {
    const response = await fetch(`${this.baseUrl}/recommendations/glove`, {
      method: 'POST',
      headers: this.headers,
      body: JSON.stringify({
        chemical: chemicalName,
        exposure_time_minutes: exposureTime,
        task_type: taskType,
      }),
    });

    if (!response.ok) {
      throw new Error(`API Error: ${response.status} ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Parser une Fiche de Données de Sécurité (FDS/SDS)
   */
  async parseSDS(file) {
    const formData = new FormData();
    formData.append('file', file);

    const response = await fetch(`${this.baseUrl}/parse/sds`, {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${this.apiKey}`,
        // Ne pas définir Content-Type pour FormData
      },
      body: formData,
    });

    if (!response.ok) {
      throw new Error(`API Error: ${response.status} ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Obtenir les informations d'un produit chimique
   */
  async getChemicalInfo(casNumber = null, name = null) {
    if (!casNumber && !name) {
      throw new Error('CAS number or name required');
    }

    const params = new URLSearchParams();
    if (casNumber) params.append('cas', casNumber);
    if (name) params.append('name', name);

    const response = await fetch(`${this.baseUrl}/chemicals?${params}`, {
      headers: this.headers,
    });

    if (!response.ok) {
      throw new Error(`API Error: ${response.status} ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Obtenir la liste de tous les matériaux de gants disponibles
   */
  async getGloveMaterials() {
    const response = await fetch(`${this.baseUrl}/gloves/materials`, {
      headers: this.headers,
    });

    if (!response.ok) {
      throw new Error(`API Error: ${response.status} ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Créer un nouveau produit chimique dans la base
   */
  async createChemical(chemicalData) {
    const response = await fetch(`${this.baseUrl}/chemicals`, {
      method: 'POST',
      headers: this.headers,
      body: JSON.stringify(chemicalData),
    });

    if (!response.ok) {
      throw new Error(`API Error: ${response.status} ${response.statusText}`);
    }

    return response.json();
  }
}

// =============================================================================
// EXEMPLES D'UTILISATION
// =============================================================================

/**
 * Exemple 1: Recommandation simple de gant
 */
async function exampleBasicRecommendation() {
  const client = new NudgeSafeClient();

  try {
    const result = await client.getGloveRecommendation('Acetone', 30, 'cleaning');

    console.log('=== Recommandation pour Acétone ===');
    console.log(`Produit: ${result.chemical.name}`);
    console.log(`CAS: ${result.chemical.cas_number}`);
    console.log('\nGants recommandés:');
    result.recommendations.recommended.forEach((glove) => {
      console.log(`  ✅ ${glove.material} - Temps percée: ${glove.breakthrough_time}min`);
    });

    console.log('\nGants NON recommandés:');
    result.recommendations.not_recommended.forEach((glove) => {
      console.log(`  ❌ ${glove.material} - Raison: ${glove.reason}`);
    });
  } catch (error) {
    console.error('❌ Erreur:', error.message);
  }
}

/**
 * Exemple 2: Parser une FDS depuis un input file
 */
async function exampleSDSParsing() {
  const client = new NudgeSafeClient();

  // Simuler sélection fichier (dans un vrai contexte, utiliser <input type="file">)
  const fileInput = document.getElementById('sds-file-input');
  const file = fileInput.files[0];

  if (!file) {
    console.error('Aucun fichier sélectionné');
    return;
  }

  try {
    const result = await client.parseSDS(file);

    console.log('=== Parsing FDS ===');
    console.log(`Produit: ${result.product_name}`);
    console.log(`CAS: ${result.cas_number}`);
    console.log('Section 8 - EPI:');
    console.log(`  Gants: ${result.ppe.gloves.join(', ')}`);
    console.log(`  Protection respiratoire: ${result.ppe.respiratory}`);
  } catch (error) {
    console.error('❌ Erreur:', error.message);
  }
}

/**
 * Exemple 3: Recherche produit chimique par CAS
 */
async function exampleChemicalLookup() {
  const client = new NudgeSafeClient();

  try {
    const result = await client.getChemicalInfo('67-64-1');

    console.log('=== Info Produit Chimique ===');
    console.log(`Nom: ${result.name}`);
    console.log(`CAS: ${result.cas_number}`);
    console.log(`Famille: ${result.family}`);
    console.log('Dangers:');
    result.hazards.forEach((hazard) => console.log(`  - ${hazard}`));
  } catch (error) {
    console.error('❌ Erreur:', error.message);
  }
}

/**
 * Exemple 4: Intégration dans une interface utilisateur
 */
function setupUI() {
  const client = new NudgeSafeClient();

  // Bouton de recherche
  document.getElementById('search-btn').addEventListener('click', async () => {
    const chemicalName = document.getElementById('chemical-input').value;
    const resultDiv = document.getElementById('result');

    try {
      resultDiv.innerHTML = '<p>Recherche en cours...</p>';

      const result = await client.getGloveRecommendation(chemicalName);

      let html = `<h3>${result.chemical.name}</h3>`;
      html += '<h4>Gants Recommandés:</h4><ul>';
      result.recommendations.recommended.forEach((glove) => {
        html += `<li>✅ ${glove.material} (${glove.breakthrough_time} min)</li>`;
      });
      html += '</ul>';

      resultDiv.innerHTML = html;
    } catch (error) {
      resultDiv.innerHTML = `<p class="error">❌ ${error.message}</p>`;
    }
  });

  // Upload FDS
  document.getElementById('sds-upload').addEventListener('change', async (e) => {
    const file = e.target.files[0];
    const statusDiv = document.getElementById('upload-status');

    if (file) {
      try {
        statusDiv.innerHTML = '<p>Parsing en cours...</p>';

        const result = await client.parseSDS(file);

        statusDiv.innerHTML = `
          <h4>FDS Parsée avec Succès</h4>
          <p><strong>Produit:</strong> ${result.product_name}</p>
          <p><strong>CAS:</strong> ${result.cas_number}</p>
          <p><strong>Gants requis:</strong> ${result.ppe.gloves.join(', ')}</p>
        `;
      } catch (error) {
        statusDiv.innerHTML = `<p class="error">❌ ${error.message}</p>`;
      }
    }
  });
}

// =============================================================================
// EXÉCUTION
// =============================================================================

// Pour Node.js
if (typeof module !== 'undefined' && module.exports) {
  module.exports = { NudgeSafeClient };
}

// Pour navigateur
if (typeof window !== 'undefined') {
  window.NudgeSafeClient = NudgeSafeClient;

  // Exécuter setupUI() quand le DOM est prêt
  if (document.readyState === 'loading') {
    document.addEventListener('DOMContentLoaded', setupUI);
  } else {
    setupUI();
  }
}

// Pour tests/démo en ligne de commande Node.js
if (typeof process !== 'undefined' && process.argv[1] === __filename) {
  console.log('NudgeSafe Hands - Exemples API JavaScript\n');
  exampleBasicRecommendation();
}
